packages/contracts/src/utils/forwarder_wrapper.ts

```diff
@@ -0,0 +1,185 @@
+        this._forwarderContract = contractInstance;
+        this._web3Wrapper = web3Wrapper;
+        this._web3Wrapper.abiDecoder.addABI(contractInstance.abi);
+    }
+    public async buyTokensAsync(
+        orders: SignedOrder[],
+        feeOrders: SignedOrder[],
+        fillAmountWei: BigNumber,
+        from: string,
+    ): Promise<TransactionReceiptWithDecodedLogs> {
+        const txOpts = {
+            from,
+            value: fillAmountWei,
+        };
+        const params = formatters.createMarketSellOrders(orders, fillAmountWei);
+        const feeParams = formatters.createMarketSellOrders(feeOrders, new BigNumber(0));
+        const txHash: string = await this._forwarderContract.buyTokens.sendTransactionAsync(
+            params.orders,
+            params.signatures,
+            feeParams.orders,
+            feeParams.signatures,
+            txOpts,
+        );
+        const tx = await this._getTxWithDecodedLogsAsync(txHash);
+        return tx;
+    }
+    public async buyTokensFeeAsync(
+        orders: SignedOrder[],
+        feeOrders: SignedOrder[],
+        fillAmountWei: BigNumber,
+        feeProportion: number,
+        feeRecipient: string,
+        from: string,
+    ): Promise<TransactionReceiptWithDecodedLogs> {
+        const txOpts = {
+            from,
+            value: fillAmountWei,
+        };
+        const params = formatters.createMarketSellOrders(orders, fillAmountWei);
+        const feeParams = formatters.createMarketSellOrders(feeOrders, new BigNumber(0));
+        const txHash: string = await this._forwarderContract.buyTokensFee.sendTransactionAsync(
+            params.orders,
+            params.signatures,
+            feeParams.orders,
+            feeParams.signatures,
+            feeProportion,
+            feeRecipient,
+            txOpts,
+        );
+        const tx = await this._getTxWithDecodedLogsAsync(txHash);
+        return tx;
+    }
+    public async buyTokensQuoteAsync(
+        orders: SignedOrder[],
+        feeOrders: SignedOrder[],
+        buyAmountWei: BigNumber,
+    ): Promise<{
+        makerAssetFilledAmount: BigNumber;
+        takerAssetFilledAmount: BigNumber;
+        makerFeePaid: BigNumber;
+        takerFeePaid: BigNumber;
+    }> {
+        const params = formatters.createMarketBuyOrders(orders, buyAmountWei);
+        const feeParams = formatters.createMarketBuyOrders(feeOrders, buyAmountWei);
+        const quote = await this._forwarderContract.buyTokensQuote.callAsync(
+            params.orders,
+            feeParams.orders,
+            buyAmountWei,
+        );
+        return quote;
+    }
+    public async buyNFTTokensAsync(
+        orders: SignedOrder[],
+        feeOrders: SignedOrder[],
+        from: string,
+    ): Promise<TransactionReceiptWithDecodedLogs> {
+        let fillAmountWei = _.reduce(
+            orders,
+            (prev: BigNumber, order: SignedOrder) => {
```

Comment:
`prev` is a misnomer here, it's not the previous value, but rather the cumulative value.